### PR TITLE
change email feature operational

### PR DESCRIPTION
**Pull request by LahlouS**

### CODE REVIEW

File: src/lib/emailRepository.ts [hunk 1]
```diff
@@ -298,13 +298,14 @@ class EmailRepository {
 		}
 	}
 
-	public async updateEmailIsSetup(userId: string, val: boolean) {
+	public updateEmailIsSetup(userId: string, val: boolean) : RunResult {
```
Comment (rutgercap) [inline src/lib/emailRepository.ts:301]: This shouldn't return run result but the result of the run mapped to a type within our domain

File: src/lib/userRepository.ts [hunk 2]
```diff
@@ -296,7 +296,24 @@ class UserRepository {
 		});
 	}
 
-	public async updateUserPswd(userId: string, password: string) {
+
+	public updateUserEmail(userId: string, email: string) : RunResult {
```
Comment (rutgercap) [inline src/lib/userRepository.ts:300]: same here

Comment (rutgercap) [reply src/lib/userRepository.ts:300]: Should also be async like the rest

File: src/lib/emailRepository.ts (hunk 1 above)
Comment (rutgercap) [reply src/lib/emailRepository.ts:301]: And async

File: src/lib/userRepository.ts [hunk 3]
```diff
@@ -296,7 +296,24 @@ class UserRepository {
...
+		try {
+			const sql = this.db.prepare<[string, string]>(
```
Comment (rutgercap) [inline src/lib/userRepository.ts:302]: This statement does not return anything so I would just return promise void
If you want you can return the amount of rows edited but I don't think it's necessary

File: src/lib/emailRepository.ts (hunk 1 above)
Comment (rutgercap) [reply src/lib/emailRepository.ts:301]: This does not actually return anything so no need to return antyhing

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		request,
+		cookies,
+		locals: { user, userRepository, emailRepository }
+	}) => {
+		const newemail = await superValidate(request, zod(newEmail));
+		
+		
+		if (!newemail) return fail(400, { newemail });
+
+		console.log('HELLO FROM THE FORM ACTION:', 'the formumlaire is valid')
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:37]: comment

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+export const load: PageServerLoad = async ({ cookies, locals }) => {
+	const user = locals.user;
+	
+	if (!user) {
+		return error(500, 'update e-mail: something went wrong on our side');
+	}
+	
+	const form = await superValidate(zod(newEmail));
+	console.log('HELLO FROM THE LOAD FUNCITON: form = ', form)
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:22]: comment

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
+import { zod } from 'sveltekit-superforms/adapters';
+import { message, superValidate } from 'sveltekit-superforms';
+import { fail, redirect, type Action } from '@sveltejs/kit';
+import { error } from '@sveltejs/kit';
+import type { PageServerLoad, Actions } from './$types';
+import { z } from 'zod';
+import { PUBLIC_BASE_URL } from '$env/static/public';
+
+const newEmail = z.object({
+	new_email: z.string().email({ message: 'Invalid email address' }),
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:10]: snake case variable names

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+		cookies,
+		locals: { user, userRepository, emailRepository }
+	}) => {
+		const newemail = await superValidate(request, zod(newEmail));
+		
+		
+		if (!newemail) return fail(400, { newemail });
+
+		console.log('HELLO FROM THE FORM ACTION:', 'the formumlaire is valid')
+		if (newemail.data.new_email !== newemail.data.confirm_email) {
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:38]: Probably makes sense to do this check on the frontend and not return the confirm to the backend at all

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+			return message(newemail, 'new email and confirmation must be the same', {
+				status: 400
+			});
+		}
+
+		if (newemail.data.new_email == user.email) {
+			return message(newemail, 'new email must be different from the previous one', {
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:45]: I wouldn't care about this tbh.
if someone want to reset it to the same email, why not

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+				status: 400
+			});
+		}
+
+
+		const update =  userRepository.updateUserEmail(user.id, newemail.data.new_email);
+		if (!update.changes) {
+			console.log('IN THE CHANGE EMAIL ACTION FUNCTION update.change = ', update.changes)
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:53]: log

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+			error(500, 'something bad happens trying to update your email')
+		}
+
+
+		const res = emailRepository.updateEmailIsSetup(user.id, false);
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:58]: Does this together with the other step of setting the email
Setting email is setup is part of the implementation of setting an email. These two functions should never be used separately so it's better to keep them together.
you can make this one private and call it from the other one

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+		if (!res.changes) {
+			console.log('IN THE CHANGE EMAIL ACTION FUNCTION EMAIL, res.changes = ', res.changes)
+			error(500, 'something bad happens trying to update your email status')
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:61]: Need try, catches for these things. These sql statements can fail with an exception and none of this is caught atm.
It's annoying but it's sadly enough part of javascript. 
you can separate it into a function so you don't have to have everything nested

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+		}
+
+		const verificationToken = emailRepository.createEmailVerificationToken(user.id, newemail.data.new_email);
+		const verificationLink = `${PUBLIC_BASE_URL}/api/email-verification/` + verificationToken;
+		console.log('IN THE CHANGE EMAIL LOAD FUNCTION : verification link = ', verificationLink);
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:66]: more log

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+		const send_details = await emailRepository.verificationLinkTo(newemail.data.new_email, verificationLink);
+		// console.log('IN THE CHANGE EMAIL LOAD FUNCTION send details from nodemailer:', send_details)
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:68]: comment

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+		}
+
+
+		const res = emailRepository.updateEmailIsSetup(user.id, false);
+		if (!res.changes) {
+			console.log('IN THE CHANGE EMAIL ACTION FUNCTION EMAIL, res.changes = ', res.changes)
+			error(500, 'something bad happens trying to update your email status')
+		}
+
+		const verificationToken = emailRepository.createEmailVerificationToken(user.id, newemail.data.new_email);
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:64]: Some of these steps probably always go together. I would make them part of the email repository interface and do them together.
At the moment most of the email repo logic is in the UI. This should all live together in the repository and the ui should just call it and return the result to the user.

Also probably most of these things should be in a Service and the service should own a repository.
A repository should just store things, at the moment it does three things:
- send emails
- handle email resetting logic
- storing tokens etc

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.svelte [hunk 15]
```diff
@@ -0,0 +1,72 @@
+<script lang="ts">
+	import type { PageData } from './$types';
+	import { superForm } from 'sveltekit-superforms';
+
+	export let data: PageData;
+
+	const { enhance, form, errors, constraints, message } = superForm(data.form);
+
+    console.log('In the front : ', form)
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.svelte:9]: comments

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.svelte [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    console.log('In the front : ', errors)
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.svelte:10]: Have a look at:
https://superforms.rocks/super-debug
This helps a lot if you're struggling to debug the superforms

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+import type { PageServerLoad, Actions } from './$types';
+import { z } from 'zod';
+import { PUBLIC_BASE_URL } from '$env/static/public';
+
+const newEmail = z.object({
+	new_email: z.string().email({ message: 'Invalid email address' }),
+	confirm_email: z.string().email({ message: 'Invalid email address' })
+});
+
+export const load: PageServerLoad = async ({ cookies, locals }) => {
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:14]: I believe cookies are not used

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+export const actions: Actions = {
+	new_email: async ({
+		request,
+		cookies,
+		locals: { user, userRepository, emailRepository }
+	}) => {
+		const newemail = await superValidate(request, zod(newEmail));
+		
+		
+		if (!newemail) return fail(400, { newemail });
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:35]: Please add braces

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+		if (newemail.data.new_email == user.email) {
+			return message(newemail, 'new email must be different from the previous one', {
+				status: 400
+			});
+		}
+
+
+		const update =  userRepository.updateUserEmail(user.id, newemail.data.new_email);
+		if (!update.changes) {
```
Comment (rutgercap) [inline src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:52]: This is not the correct check.
If you want to check if data is changed in the database, write a test to see if the function works
If you need to know how many rows changed, return the amount of rows changed (it's typically considered incorrect to error for no rows changed on an UPDATE or DELETE query)
Here you should just catch any errors that happened

**Review by rutgercap — COMMENTED** [review]
Mostly cleaning up comments and some interface updates.
Generally: 
- remove logs before pushing to main
- remove unused variables
- If functions are always called together (setting an email and then setting email_is_set) then combine the logic in the service/repository/whatever instead of the UI
- services etc are async
- Services / repositories should not return implementation details, but something useful in our domain (RunResult is implementation detail from sqlite)
- if an

File: src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts (hunk 19 above)
Comment (LahlouS) [reply src/routes/(authenticated)/profile/[user_id]/edit-profile/change-email/+page.server.ts:52]: I have removed the ugly check and made test functions in respective test repository

**Pull request closed without merging**